core/address/codec.go

```diff
@@ -2,8 +2,10 @@ package address
 
 // Codec defines an interface to convert addresses from and to string/bytes.
 type Codec interface {
-	// StringToBytes decodes text to bytes
+	// GetAddressPrefix returns the address prefix (if any).
+	GetAddressPrefix() string
```

Comment:
We can't assume the codec is always bech32. I'd rather find another way to deal with this

Reply:
Yes (https://github.com/cosmos/cosmos-sdk/pull/15913#discussion_r1174609290), so do you find this naming not fitting if the address has no prefix?

I do not know if you've seen the autocli PR but this come handy there.

Reply:
Is that fine for autocli to depends directly auth with then?
That would solve the issue I have there.

Reply:
since autocli is built off of grpc why not use https://github.com/cosmos/cosmos-sdk/blob/be081b10515fce26aac2dbfd9ea0f4a8052e0bd1/proto/cosmos/auth/v1beta1/query.proto#L64. This way autocli/hubl will only work with bech32 addresses for now, but there can be a fallback or a way for a user to load their own

Reply:
It is almost what it already uses there, and it works fine for hubl, which will build the command after having queried an endpoint. However, for normal commands, like `simd`, `gaiad`, etc.. When we enrich the root command, we have no endpoint to query. The information is however available, and imho this was the cleanest way to get it.

We could however:

- Add an argument in EnhanceRootCommand for the bech32 prefix (easiest but most ugly because for hubl we'll need to pass an empty string)
- Inject address codec and add api to get address prefix (this was my preferred solution)
- Inject auth configuration and use `config.Bech32Prefix` for the address prefix, which is fine for app wiring app, but it makes it weird when you have to construct AppOptions manually
- Inject auth, then we have a hard dependency on auth for autocli

I am open to other suggestions however.

Reply:
Why don't we just change the flag.Builder argument to take address.Codec instead of bech32 prefix. I think that would resolve this without needing to change address.Codec. Also if we do need the actual prefix let's use an extension interface instead of changing address.Codec

Reply:
I didn't think of the problem this way. This would indeed work too, and only complexify the flag builder
I will close this then and re-open it shortly with only the clean-ups that were done here.